Rendering options for footnotes in HTML

funkyfuture opened the pull request:
Subject: Rendering options for footnotes in HTML

### Feature or Bugfix
- Feature

### Purpose
- introduce options to format rendered footnotes by the html builder

### Detail
- option to display all footnotes at the bottom of a document
  - optionally introduced by a localized heading or rubric
- option to expand footnote references with a `title` attribute containing the footnote's text

### Relates
- #3974

funkyfuture committed:
Fixes a type: "footenotes"

funkyfuture committed:
Adds 'html_footnotes_options', allows grouping of footnotes at bottom

funkyfuture committed:
html: Adds an option that display footnotes as tooltip

Therefore also a transform is added that assigns a footnote's text to
its reference nodes.

Closes #3974.

tk0miya commented on CHANGES:
```diff
@@ -93,7 +95,6 @@ Features added
 --------------
 
 * #4107: Make searchtools.js compatible with pre-Sphinx1.5 templates
-* #4112: Don't override the smart_quotes setting if it was already set
```
It seems some CHANGES entry has been lost. Could you update this please?

tk0miya commented on CHANGES:
```diff
@@ -103,7 +104,6 @@ Bugs fixed
 * #3987: Changing theme from alabaster causes HTML build to fail
 * #4096: C++, don't crash when using the wrong role type. Thanks to mitya57.
 * #4070, #4111: crashes when the warning message contains format strings (again)
-* #4108: Search word highlighting breaks SVG images
```
This also.

tk0miya commented on doc/config.rst:
```diff
@@ -1092,10 +1092,21 @@ that use Sphinx's HTMLWriter class.
 .. confval:: html_experimental_html5_writer
 
-   Output is processed with HTML5 writer.  This feature needs docutils 0.13 or newer.  Default is ``False``.
+   Output is processed with HTML5 writer. This feature needs docutils 0.13 or
```
Our document uses double spaces after sentence. Please don't remove it.

tk0miya commented on doc/config.rst:
```diff
@@ -1092,10 +1092,21 @@ that use Sphinx's HTMLWriter class.
+   at the bottom of a document. With the additional ``rubric`` or ``title`` a
+   rubric directive respectively a title are prepended.
+   The text of a footnote will be rendered as tooltip when ``tooltip`` is in
+   the configuration value.
```
What is default value of this setting?

funkyfuture replied:
there's none.

tk0miya replied:
I prefer the default setting is also named and described here.

tk0miya commented on sphinx/writers/html.py:
```diff
@@ -863,6 +866,83 @@ def visit_math(self, node, math_env=''):
+
+        if 'bottom' in self.builder.config.html_footnotes_options:
+            self.collected_footnotes.append(node)
+            raise nodes.SkipChildren
```
`nodes.SkipNode` is better for this case. then you don't care depart_footnote.

tk0miya commented on sphinx/writers/html.py:
```diff
@@ -863,6 +866,83 @@ def visit_math(self, node, math_env=''):
+            for ref in self.footnote_refs.pop(_id):
+                self.body[ref] = self.body[ref].replace(
+                    '>', u(' title="{}">').format(
+                        node[1].astext().replace('\n', ' ')))
```
I feel this is very fragile. I think it is better to collect footnote titles before translation phase. It enables to generate HTML tags for footnote_reference simply.

funkyfuture replied:
yeah, i was also surprised that it works. but there are certainly corner cases where it would break. where should i put the collection routine?

tk0miya commented on sphinx/writers/html.py:
```diff
@@ -863,6 +866,83 @@ def visit_math(self, node, math_env=''):
+        BaseTranslator.depart_document(self, node)
+
+    def build_footnotes(self):
+        footnotes = self.document.copy()
```
Why copying document?

funkyfuture replied:
iirc the builder wanted at least a `setting` property on the root node. the `copy` method seemed to intended for this kind of use.

tk0miya commented on sphinx/writers/html.py:
```diff
@@ -863,6 +866,83 @@ def visit_math(self, node, math_env=''):
+
+    def depart_rubric(self, node):
+        HTMLTranslator.depart_rubric(self, node)
+        self._start_footnotes_table()
```
Does this translator handle rubric which users written?

funkyfuture replied:
the user should either define `rubric` in the new option or put a `rubric` directive at the end of a document where s_he knows that there are footnotes.

tk0miya commented on tests/roots/test-footnotes-html/conf.py:
```diff
@@ -0,0 +1,7 @@
+# -*- coding: utf-8 -*-
```
Please rename this to more suitable directory name.

funkyfuture replied:
i'd appreciate a proposal.

tk0miya commented:
We have to discuss the policy of sphinx.
* Is this suitable for core feature?
* Is it okay for HTML builder only?
* any more options?

funkyfuture commented:
> Is it okay for HTML builder only?

the LaTeX builder puts all footnotes at the bottom of a page. i don't know how suitable this would be for other builders.

funkyfuture commented:
i added a transform that assigns the footnote's text to the referring note before the document is written.

funkyfuture commented:
@tk0miya if there's still stuff that i need to resolve, can you give me directions?

tk0miya commented on sphinx/transforms/__init__.py:
```diff
@@ -304,6 +304,27 @@ def apply(self):
+            refid = node.get('refid')
+            if refid is None:
+                continue
+            node['footnote_text'] = texts[refid]
```
I think `footnote_` prefix is not needed here. The `footnote_reference` node already knows it treats about footnotes. How about renaming it to `title` simply?

funkyfuture replied:
with the naming i wanted to make sure that it's clear that this data comes from the related node, and i'd like to keep it for that purpose.

tk0miya commented on sphinx/transforms/__init__.py:
```diff
@@ -304,6 +304,27 @@ def apply(self):
+        for node in self.document.traverse(nodes.footnote):
+            _text = node[1].astext().replace('\n', ' ')
+            for _id in node['ids']:
+                texts[_id] = _text
```
Sphinx usually uses non-underscored variables even for local variable . So you don't need to prepend it.

funkyfuture replied:
the `_id` is prefixed to not override the builtin `id`. the other i change.

tk0miya commented on doc/config.rst:
```diff
@@ -1116,10 +1116,21 @@ that use Sphinx's HTMLWriter class.
 
    .. versionadded:: 1.6
 
+.. confval:: html_footnotes_options
```
I guess this is a list option. Right? If true, please mention it.

tk0miya commented on sphinx/writers/html.py:
```diff
@@ -864,6 +865,70 @@ def visit_math(self, node, math_env=''):
+        if 'tooltip' in self.builder.config.html_footnotes_options \
+                and 'footnote_text' in node:
+            x = self.body[-1]
+            x = x.replace('>', u(' title="{}">').format(node['footnote_text']))
```
I think fully overriding is better for this case. We must not expect the behavior of super class. It might be changed in future.

funkyfuture replied:
i could copy that code. but i think it's sufficient that the resulting html is checked in the tests and any changes in the base class would be noticed with these failing.

tk0miya commented on sphinx/writers/html.py:
```diff
@@ -864,6 +865,70 @@ def visit_math(self, node, math_env=''):
+    def build_footnotes(self):
+        footnotes = self.document.copy()
+        footnotes.extend(self.collected_footnotes)
+        visitor = _BottomFootnotesTranslator(self.builder, footnotes)
```
I think the responsibility of document transformation should be on `Transform` module. The responsibility of `Translator` is a representation. In other words, `Translator` class goes to map each node to HTML tags.
So these modification should move to `Transform` module.

For example, latex writer uses `ShowUrlsTransform` (see `sphinx/writers/latex.py`).

funkyfuture replied:
that looks reasonable. but that would mean that for each footnote an indidual `table` element would be created, which actually is the current behaviour. i can think of addressing this by introducing a new node type `footnotes` which would be a container for a sequence of footnotes (at some point in the document) and can be used by writers (to construct a single table element in the html case). should i take this approach?

funkyfuture replied:
@tk0miya i'd appreciate a short feedback on this point before i begin working on it.

tk0miya commented on sphinx/builders/html.py:
```diff
@@ -1389,6 +1389,7 @@ def setup(app):
     app.add_config_value('html_search_scorer', '', None)
     app.add_config_value('html_scaled_image_link', True, 'html')
     app.add_config_value('html_experimental_html5_writer', None, 'html')
+    app.add_config_value('html_footnotes_options', '', 'html')
```
Please add fourth argument using `sphinx.config.ENUM` like following:
```
    app.add_config_value('latex_engine', default_latex_engine, None,
                         ENUM('pdflatex', 'xelatex', 'lualatex', 'platex'))
```
It detects wrong values on running..

funkyfuture replied:
i don't think that `ENUM` is feasible to validate the values as multiple values can be set as flags. e.g. these are valid: `'bottom,tooltip'`, `['bottom', 'rubric']`.

tk0miya commented on doc/config.rst:
```diff
@@ -1116,10 +1116,21 @@ that use Sphinx's HTMLWriter class.
+   at the bottom of a document.  With the additional ``rubric`` or ``title`` a
+   rubric directive respectively a title are prepended.
+   The text of a footnote will be rendered as tooltip when ``tooltip`` is in
+   the configuration value.
```
I'd like to improve `bottom` option to support other builders. Personally, I've usually placed footnotes at the end of each sections. So it would be nice if this also supports it. Are you interested in?

funkyfuture replied:
well, i don't have the time to implement that now, but i'd change the current option's name to `page-bottom` (or should it be `content-bottom`?) for that extension.

funkyfuture commented:
@tk0miya, this is stalled for a while now and it becomes harder for me to get a grip on it again. iirc, there are two obstacles i need to be cleared before i can update the pr accordingly:

shall i ['invent' the node type `footnotes`](https://github.com/sphinx-doc/sphinx/pull/4127#discussion_r155931400). that'd imply an extension of `docutils`' parser or an additional transform. and i have no clue how other writers would deal with that.

how should the configuration values be structured? atm the options `bottom`, `rubric`, `title` and `tooltip` can be provided either in a container with strings (e.g. `['bottom', 'title']`) or as a single string (e.g. `'bottom, title'`). `rubric` and `title` are only considered if `bottom` is set and are exclusive. otherwise there are no dependencies between these options.

a minor issue that can be changed easily: should the `bottom` option be renamed to `page-bottom`? as an extension to place footnotes at the end of a section is also a possible future enhancement.

be aware that i might have gotten some categories wrong here, but i still haven't memorized (or even understood) all the mechanics that make the magic happen.

schettino72 committed:
quickstart: set conf.py pygments_style value to None

Currently quickstart sets a value for `pygments_style`,
this is not desirable because it overrides the theme's default `pygments_style`.
It should be `None` so theme`s value is used by default.

tk0miya committed:
Fix #5348: download reference to remote file is not displayed

tk0miya committed:
Merge branch '1.8' into 5348_support_remote_download_file

tk0miya committed:
Merge pull request #5349 from tk0miya/5348_support_remote_download_file

Fix #5348: download reference to remote file is not displayed

tk0miya committed:
Merge pull request #5282 from schettino72/quickstart-pygments-style-none

quickstart: set conf.py pygments_style value to None

tk0miya committed:
Add testcase for pygments

tk0miya committed:
Update CHANGES for #5282

tk0miya committed:
Merge pull request #5371 from tk0miya/5282_testcase_for_pygments

testcases and CHANGES for #5282

tk0miya committed:
Fix CHANGES
